pkg/conversion/mapi2capi/machine.go

```diff
@@ -79,3 +87,363 @@ func fromMAPIMachineToCAPIMachine(mapiMachine *mapiv1beta1.Machine, apiVersion,
+}
+
+// convertMAPIMachineAddressesToCAPI converts MAPI machine addresses to CAPI format.
+func convertMAPIMachineAddressesToCAPI(mapiAddresses []corev1.NodeAddress) (clusterv1.MachineAddresses, field.ErrorList) {
+	if mapiAddresses == nil {
+		return nil, nil
+	}
+
+	errs := field.ErrorList{}
+	capiAddresses := make(clusterv1.MachineAddresses, 0, len(mapiAddresses))
+
+	// Addresses are slightly different between MAPI/CAPI.
+	// In CAPI the address type can be: Hostname, ExternalIP, InternalIP, ExternalDNS or InternalDNS
+	// In MAPI the address type can be: Hostname, ExternalIP, InternalIP (missing ExternalDNS and InternalDNS)
+	// This is fine when going from MAPI to CAPI, but needs to be handled when going from CAPI to MAPI.
+	for _, addr := range mapiAddresses {
+		var t clusterv1.MachineAddressType
+
+		switch addr.Type {
+		case corev1.NodeHostName:
+			t = clusterv1.MachineHostName
+		case corev1.NodeExternalIP:
+			t = clusterv1.MachineExternalIP
+		case corev1.NodeInternalIP:
+			t = clusterv1.MachineInternalIP
+		case corev1.NodeExternalDNS:
+			t = clusterv1.MachineExternalDNS
+		case corev1.NodeInternalDNS:
+			t = clusterv1.MachineInternalDNS
+		default:
+			errs = append(errs, field.Invalid(field.NewPath("status", "addresses"), string(addr.Type), string(addr.Type)+" unrecognized address type"))
+		}
+
+		capiAddresses = append(capiAddresses, clusterv1.MachineAddress{
+			Type:    t,
+			Address: addr.Address,
+		})
+	}
+
+	return capiAddresses, errs
+}
+
+// convertMAPIMachinePhaseToCAPI converts MAPI machine phase to CAPI format.
+func convertMAPIMachinePhaseToCAPI(mapiPhase *string) string {
+	// Phase is slightly different between MAPI/CAPI.
+	// In CAPI can be one of: Pending;Provisioning;Provisioned;Running;Deleting;Deleted;Failed;Unknown
+	// In MAPI can be one of: Provisioning;Provisioned;Running;Deleting;Failed (missing Pending,Unknown)
+	// This is fine when going from MAPI to CAPI, but needs to be handled when going from CAPI to MAPI.
+	// MAPI and CAPI phases are compatible, but we need to handle the pointer vs string difference
+	return ptr.Deref(mapiPhase, "")
+}
+
+// convertMAPIMachineErrorReasonToCAPIFailureReason converts MAPI MachineStatusError to CAPI MachineStatusError.
+func convertMAPIMachineErrorReasonToCAPIFailureReason(mapiErrorReason *mapiv1beta1.MachineStatusError) *capierrors.MachineStatusError {
+	if mapiErrorReason == nil {
+		return nil
+	}
+
+	return ptr.To(capierrors.MachineStatusError(*mapiErrorReason))
+}
+
+// convertMAPIMachineErrorMessageToCAPIFailureMessage converts MAPI MachineStatusError to CAPI MachineStatusError.
+func convertMAPIMachineErrorMessageToCAPIFailureMessage(mapiErrorMessage *string) *string {
+	if mapiErrorMessage == nil {
+		return nil
+	}
+
+	return ptr.To(*mapiErrorMessage)
+}
+
+// deriveCAPIBootstrapReadyFromMAPI derives the CAPI BootstrapReady field from MAPI machine state.
+func deriveCAPIBootstrapReadyFromMAPI(mapiMachine *mapiv1beta1.Machine) bool {
```

Comment:
optional/nit: I'd consider the more specialized funcs to be at the bottom, helps readability.

So e.g. `convertMAPIMachineConditionsToCAPIMachineConditions` should be above these `derive` funcs.